### PR TITLE
Rework the ffi key provider and tweak key loading

### PR DESCRIPTION
**Pull request by dewyatt**
The main thing here is changing how the key provider works in the FFI. This changes it to just be a signal to inform the application that it should load the key via the ffi methods.

I also changed key loading a bit to use the key provider where it seemed to make sense.
This allows, for example, loading a subkey for which the primary wasn't loaded yet. The G10 store also now uses the key provider to load the pubkey, instead of having that confusing pubring parameter in there.

(I did some last minute changes here so I'll probably want to review this myself tomorrow too.)

### CODE REVIEW

**Comment by dewyatt** [comment]
Just noticed I forgot to commit a test I added so I'll get that in shortly too.

**Comment by ronaldtse** [comment]
@ni4 could you help review / approve / merge this? Thanks!

File: src/lib/key-provider.c [hunk 1]
```diff
@@ -24,55 +24,114 @@
...
+    case PGP_KEY_SEARCH_FINGERPRINT:
+        return (key->fingerprint.length == search->by.fingerprint.length) &&
+               !memcmp(key->fingerprint.fingerprint,
+                       search->by.fingerprint.fingerprint,
+                       key->fingerprint.length);
+    case PGP_KEY_SEARCH_GRIP:
+        return !memcmp(key->grip, search->by.grip, PGP_FINGERPRINT_SIZE);
+    case PGP_KEY_SEARCH_USERID: {
+        for (unsigned i = 0; i < key->uidc; i++) {
+            if (!strcmp((char *) key->uids[i], search->by.userid)) {
```
Comment (ni4) [inline src/lib/key-provider.c]: Maybe use strstr/strcasestr here? Or do we need exact match here?

Comment (dewyatt) [reply src/lib/key-provider.c]: My feeling is it's best to go for an exact match within the key provider itself. An application using the FFI can always use the id iterator to do fancier matces like case-insensitive/regex userid, or partial keyid, and so on. (Or we could implement an FFI search function with a callback to check custom criteria, etc.)

**Comment by dewyatt** [comment]
It looks like I left some debugging stuff in here, so let me get rid of that...

File: src/lib/key-provider.c (hunk 1 above)
Comment (ronaldtse) [reply src/lib/key-provider.c]: @ni4 if it’s OK feel free to merge it. Thanks!

**Pull request merged**